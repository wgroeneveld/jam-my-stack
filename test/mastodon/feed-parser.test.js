
describe("mastodon feed parser tests", () => {
	const fs = require('fs');
	const fsp = require('fs').promises;
	const { rmdir } = require('./../utils')

	const frontMatterParser = require('parser-front-matter');

	jest.mock('got');

	const { parseMastoFeed } = require('../../src/mastodon/feed-parser')
	const dumpdir = `${__dirname}/dump`

	beforeEach(() => {
		if(fs.existsSync(dumpdir)) {
			rmdir(dumpdir)
		}
		fs.mkdirSync(dumpdir)
	});

	describe("ignore replies option", () => {
		test("ignore replies if ignoreReplies is set to true", async () => {
			await parseMastoFeed({
				url: "masto-feed-with-replies",
				notesdir: dumpdir,
				ignoreReplies: true
			})

			dir = await fsp.readdir(`${dumpdir}/2021/03`, { withFileTypes: true })
			expect(dir.length).toBe(1)		
		})
		test("does not ignore replies if ignoreReplies is set to false", async () => {
			await parseMastoFeed({
				url: "masto-feed-with-replies",
				notesdir: dumpdir,
				ignoreReplies: false
			})

			dir = await fsp.readdir(`${dumpdir}/2021/03`, { withFileTypes: true })
			expect(dir.length).toBe(2)		
		})
	})

	test("parse embedded images", async () => {
		await parseMastoFeed({
			url: "masto-feed-images",
			notesdir: dumpdir,
			utcOffset: 1
		})

		const actualMd = (await fsp.readFile(`${dumpdir}/2021/03/14h17m41s53.md`)).toString()
		expect(actualMd).toMatchSnapshot()
	})

	test("parse prepends double quotes with backlash to escape in frontmatter", async () => {
		await parseMastoFeed({
			url: "masto-feed-quote",
			notesdir: dumpdir,
			utcOffset: 0
		})

		const actualMd = await fsp.readFile(`${dumpdir}/2021/03/02h16m18s46.md`)
		
		const md = frontMatterParser.parseSync(actualMd.toString())
		expect(md.data.title).toBe("\"wow this sucks\" with quotes")
	})

	test("parse trims title according to config and adds three dots", async () => {
		await parseMastoFeed({
			url: "masto-feed-sample",
			notesdir: dumpdir,
			utcOffset: 0,
			titleCount: 5,
			titlePrefix: "Note: "
		})

		const actualMd = await fsp.readFile(`${dumpdir}/2021/03/02h16m18s46.md`)
		
		const md = frontMatterParser.parseSync(actualMd.toString())
		expect(md.data.title).toBe("Note: @Stam...")
	})

	test("parse does not trim if titleCount > title length and does not add three dots", async () => {
		await parseMastoFeed({
			url: "masto-feed-sample",
			notesdir: dumpdir,
			utcOffset: 0,
			titleCount: 5000
		})

		const actualMd = await fsp.readFile(`${dumpdir}/2021/03/02h16m18s46.md`)
		
		const md = frontMatterParser.parseSync(actualMd.toString())
		expect(md.data.title).toBe("@StampedingLonghorn I tried to chase him away, but you know how that turned out... ðŸ˜¼ There's ...")
	})

	test("parse creates separate notes in each month subdir", async () => {
		await parseMastoFeed({
			url: "masto-feed-sample",
			notesdir: dumpdir
		})

		let dir = await fsp.readdir(`${dumpdir}/2021/02`, { withFileTypes: true })
		expect(dir.length).toBe(8)
		dir = await fsp.readdir(`${dumpdir}/2021/03`, { withFileTypes: true })
		expect(dir.length).toBe(4)
	})

	test("parse creates correct MD structure", async () => {
		await parseMastoFeed({
			url: "masto-feed-sample",
			notesdir: dumpdir,
			utcOffset: 0,
			titleCount: 5000			
		})

		const actualMd = (await fsp.readFile(`${dumpdir}/2021/03/01h19m03s35.md`)).toString()
		expect(actualMd).toMatchSnapshot()
	})

	test("parse creates MD with context if in-reply-to", async () => {
		//https://aus.social/users/aussocialadmin/statuses/105817435308293091
		await parseMastoFeed({
			url: "masto-feed-sample",
			notesdir: dumpdir,
			utcOffset: 0,
			titleCount: 5000			
		})

		const actualMd = await fsp.readFile(`${dumpdir}/2021/03/02h16m18s46.md`)
		const expectedReplyTo = "https://social.linux.pizza/users/StampedingLonghorn/statuses/105821099684887793"

		const md = frontMatterParser.parseSync(actualMd.toString())
		expect(md.data.context).toBe(expectedReplyTo)
	})
})
