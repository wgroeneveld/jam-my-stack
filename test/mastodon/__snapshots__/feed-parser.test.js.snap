// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`mastodon feed parser tests parse creates correct MD structure 1`] = `
"---
source: \\"https://chat.brainbaking.com/objects/77a3ecfb-47e1-4d7a-a24a-8b779d80a8ac\\"
context: \\"\\"
title: \\"I pulled the Google plug and installed LineageOS: https://brainbaking.com/post/2021/03/getting-ri...\\"
date: \\"2021-03-01T19:03:35\\"
---

I pulled the Google plug and installed LineageOS: <a href=\\"https://brainbaking.com/post/2021/03/getting-rid-of-tracking-using-lineageos/\\" rel=\\"ugc\\">https://brainbaking.com/post/2021/03/getting-rid-of-tracking-using-lineageos/</a> Very impressed so far! Also rely on my own CalDAV server to replace GCalendar. Any others here running <a class=\\"hashtag\\" data-tag=\\"lineageos\\" href=\\"https://chat.brainbaking.com/tag/lineageos\\" rel=\\"tag ugc\\">#lineageos</a> for privacy reasons?
"
`;

exports[`mastodon feed parser tests parse embedded images 1`] = `
"---
source: \\"https://chat.brainbaking.com/objects/a51e13ce-d618-4602-84f7-f398126510ff\\"
context: \\"\\"
title: \\"Enjoyed an afternoon of oldskool Diablo II on the ...\\"
date: \\"2021-03-14T17:41:53\\"
---

Enjoyed an afternoon of oldskool Diablo II on the Europebattle servers. We did a few Mephisto runs, managed to hit Hell, and I re-converetd my druid into a windy one. Good times!

<div class=\\"flex\\">

<div>
<a class=\\"lbox\\" href=\\"https://chat.brainbaking.com/media/6f8b72ca-9bfb-460b-9609-c4298a8cab2b/EuropeBattle%202021-03-14%2016-20-36-87.jpg\\">
<img loading=\\"lazy\\" src=\\"https://chat.brainbaking.com/media/6f8b72ca-9bfb-460b-9609-c4298a8cab2b/EuropeBattle%202021-03-14%2016-20-36-87.jpg\\" alt=\\"Enclosed Toot image\\">
</a>
</div>

<div>
<a class=\\"lbox\\" href=\\"https://chat.brainbaking.com/media/3dbcb044-2acc-4ace-a4f6-37ce94c3f2b1/EuropeBattle%202021-03-14%2015-35-01-56.jpg\\">
<img loading=\\"lazy\\" src=\\"https://chat.brainbaking.com/media/3dbcb044-2acc-4ace-a4f6-37ce94c3f2b1/EuropeBattle%202021-03-14%2015-35-01-56.jpg\\" alt=\\"Enclosed Toot image\\">
</a>
</div>

</div>"
`;
